lls/cache.c

```diff
@@ -0,0 +1,468 @@
+	}
+}
+
+static void
+lls_process_hold(struct lls_config *lls_conf, struct lls_hold *hold)
+{
+	struct lls_cache *cache = hold->cache;
+	struct lls_record *record;
+	int ret = rte_hash_lookup(cache->hash, hold->ip_be);
+
+	if (ret == -ENOENT) {
+		ret = lls_add_record(cache, hold->ip_be);
+		if (ret < 0)
+			return;
+
+		record = &hold->cache->records[ret];
+		record->map.stale = true;
+		memcpy(record->map.ip_be, hold->ip_be, cache->key_len);
+		record->ts = time(NULL);
+		RTE_ASSERT(record->ts >= 0);
+		record->holds[0] = *hold;
+		record->num_holds = 1;
+
+		/* Try to resolve record using broadcast. */
+		lls_send_request(lls_conf, cache, hold->ip_be, NULL);
+
+		if (lls_conf->debug)
+			lls_cache_dump(cache);
+		return;
+	} else if (unlikely(ret == -EINVAL)) {
+		char ip_buf[cache->key_str_len];
+		char *ip_str;
+		ip_str = cache->ip_str(cache, hold->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, could not get %s map; hold failed\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	RTE_ASSERT(ret >= 0);
+	record = &hold->cache->records[ret];
+
+	if (!record->map.stale) {
+		int call_again = false;
+		/* Alert requester this map is ready. */
+		hold->cb((const struct lls_map *)&record->map, hold->arg,
+			LLS_REPLY_RESOLUTION, &call_again);
+		if (!call_again)
+			return;
+	}
+	record->holds[record->num_holds++] = *hold;
+
+	if (lls_conf->debug)
+		lls_cache_dump(cache);
+}
+
+static void
+lls_process_put(struct lls_config *lls_conf, struct lls_put *put)
+{
+	struct lls_cache *cache = put->cache;
+	struct lls_record *record;
+	unsigned int i;
+	int ret = rte_hash_lookup(cache->hash, put->ip_be);
+
+	if (ret == -ENOENT) {
+		/*
+		 * Not necessarily an error: the block may have indicated
+		 * it did not want its callback to be called again, and
+		 * all holds have been released on that entry.
+		 */
+		return;
+	} else if (unlikely(ret == -EINVAL)) {
+		char ip_buf[cache->key_str_len];
+		char *ip_str = cache->ip_str(cache, put->ip_be, ip_buf,
+			cache->key_str_len);
+		RTE_LOG(ERR, HASH,
+			"Invalid params, could not get %s map; put failed\n",
+			ip_str == NULL ? cache->name : ip_str);
+		return;
+	}
+
+	RTE_ASSERT(ret >= 0);
+	record = &put->cache->records[ret];
+
+	for (i = 0; i < record->num_holds; i++) {
+		if (put->lcore_id == record->holds[i].lcore_id)
+			break;
+	}
+
+	/* Requesting lcore not found in holds. */
+	if (i == record->num_holds)
+		return;
+
+	/*
+	 * Alert the requester that its hold will be removed, so it
+	 * may free any state that is keeping track of that hold.
+	 *
+	 * Technically the hold will be removed in the step
+	 * below, but alerting the requester first removes the need
+	 * to copy the hold into a temporary variable, remove
+	 * the hold from record->holds, and then alert the
+	 * requester using the temporary variable. This is OK since
+	 * there's only one writer.
+	 */
+	record->holds[i].cb((const struct lls_map *)&record->map,
```

Comment:
Is the cast `(const struct lls_map *)` really needed here? If not, please search the code for other occurrences.